internal/pipeline/pipeline.go

```diff
@@ -79,14 +81,63 @@ func (p *Pipeline) String() string {
 	return fmt.Sprintf("%+v", *p)
 }
 
+// Validate will validate the Pipeline `p`. This function will validate
+// fields 'apiVersion', 'kind', 'sources', 'generators', 'transformers'
+// and 'validators'.
+//
+// 'sources' are valid if and only if every source is:
+// 1. a relative path to local package OR
+// 2. '.' OR
+// 3. './*'
+//
+// 'generators', 'transformers' and 'validators' share same schema and
+// they are valid if all functions in them are ALL valid.
+func (p *Pipeline) Validate() error {
+	if p.APIVersion != kptAPIVersion {
+		return fmt.Errorf("apiVersion %s is not valid, should be %s",
+			p.APIVersion, kptAPIVersion)
+	}
+	if p.Kind != pipelineKind {
+		return fmt.Errorf("kind %s is not valid, should be %s",
+			p.Kind, pipelineKind)
+	}
+	for _, s := range p.Sources {
+		if s == "." || s == "./*" {
+			continue
+		}
+		if filepath.IsAbs(s) {
```

Comment:
Need more validation for source. Should extract this in its own function.

These paths are slash-separated OS-agnostic paths so 'filepath' package (OS-specific) is not appropriate. Look at the 'path' package instead, keep in mind we define special characters like './*'.

Need to check it's a) Slash delimited b) Relative c) Clean 

Examples of bad sources:

a\b\c
c:\a\b\c\
/home/frankf
../a/../b

There are probably other corner cases that needs to be considered. Shouldn't assume kpt is run on a Unix system.

Reply:
How does kpt handle paths on non-Unix system?

Reply:
How do you define `Clean`? if it's `path.Clean()` then `./base` is not a clean path.

Reply:
Re How does kpt handle paths on non-Unix system?

Depends on which level you're talking about. Arguments to CLI or OS-specific. In this context, we're talking about an OS-agnostic, portable format for Pipeline.sources. 

How do you define Clean?

We are defining our own format in Pipeline.sources which may have nothing to do with Golang or its std libraries. We can decide how strict we want the validation to be.
For example, which of the following should we accept:

`base`, `./base`, `./base/`, `base/`, `base1/.../base2`?

There's a trade-off here:
1) Strict validation ensures ensures uniformity and consistency.
2) Strict validation may be unnecessary pedantic or "dumb". For example rejecting, `base/` even though it's clear
what the user intent was.
 
The most pragmatic (easiest to implement) is:
1. Treat `./*` as a special case.
2. Accept any other relative path (not strict) and run Clean on it. All cases above are acceptable.

Also note:
- Reason we chose `./base` was because it makes it obvious this is a relative to the current directory. With `base`, there may be an ambiguity whether package `base` lives in the filesystem.

Thoughts?

Reply:
+1 for the pragmatic way.